### PR TITLE
Add ColorGradientFilter

### PR DESCRIPTION
**Pull request by vikpe**
An attempt at creating a generic color gradient filter.

[color_gradient_filter.webm](https://user-images.githubusercontent.com/1616817/209396849-7114d3b7-50c8-4965-a640-7a96f6ff497e.webm)

**Options**
* Colorstops[] (`offset`, `color`, `alpha`)
* Type: `linear`, `radial`, `conic`
* `alpha`
* `angle`
* `max colors` (limit color palette)

**Some ideas for improvement**
* Use anti-aliasing
* Allow to offset the gradient (the entire result)
* Allow to scale the gradient
* Optimize fragment shader (it's very likely suboptimal)
* More gradient types (?)

Feedback is very much welcome!

### CODE REVIEW

**Comment by bigtimebuddy** [comment]
Another cool feature that would make this super useful is to have the ability to convert from CSS. Since there are a lot of tools that can generate CSS gradients (eg Figma, Sketch), it would make it more flexible. For example:

```
linear-gradient(90deg, rgba(2,0,36,1) 0%, rgba(9,9,121,1) 35%, rgba(0,212,255,1) 100%);
```

Another idea would be to support (translate) from CanvasGradient object.

**Comment by vikpe** [comment]
Good suggestions, will fix!

**Comment by bigtimebuddy** [comment]
I think if you add gradient-parser as devDependency it should bundle it. Then you don't have to worry about user importing.

**Comment by vikpe** [comment]
> I think if you add gradient-parser as devDependency it should bundle it. Then you don't have to worry about user importing.

You mean that it will work in final rollup build? What about during development?

```sh
rollup v2.79.1
bundles filters/color-gradient/src/index.ts → filters/color-gradient/dist/filter-color-gradient.js, filters/color-gradient/dist/filter-color-gradient.mjs...
[!] Error: 'default' is not exported by node_modules/gradient-parser/build/node.js, imported by filters/color-gradient/src/ColorGradientFilter.ts
```

**Comment by vikpe** [comment]
btw, dependencies are lagging pretty far behind:

image

**Comment by bigtimebuddy** [comment]
Lets do a dependency update in another PR. I'm not concerned about those since what we're doing here is pretty straightforward.

I'll look at the gradient-parser thing and get back to you.

**Comment by bigtimebuddy** [comment]
Okay, I added `@rollup/plugin-commonjs` to be able to import gradient-parser. Should be good now!

**Comment by vikpe** [comment]
It works, thanks! 

However the result from `gradient-parser` is complex and can't be applied to the shader directly, I still need to do a lot of conversions (offsets, colors etc).

e.g. `linear-gradient(red, green, blue 80%, red, green 90%, blue)` equals offsets `0.0, 0.4, 0.8, 0.85, 0.9, 1.0`.

File: filters/color-gradient/src/ColorGradientFilter.ts [hunk 1]
```diff
@@ -54,6 +54,14 @@ class ColorGradientFilter extends Filter
         maxColors: 0,
     };
 
+    public static fromCssGradient(cssGradient: string, options?: Partial<ColorGradientFilterOptions>) : ColorGradientFilter
```
Comment (bigtimebuddy) [inline filters/color-gradient/src/ColorGradientFilter.ts:57]: The static from method feels a little clunky and anti-pattern to me. No other filters do this. My suggestion is to have different option shapes a user could pass.

```ts
type ColorGradientFilterOptions = {
    type: number;
    stops: ColorStop[];
    angle: number;
    alpha: number;
    maxColors: number;
} | {
    css: string;
    alpha: number;
    maxColors: number;
};
```
Then in the constructor you could parse the CSS like you're doing here.

**Comment by vikpe** [comment]
image

now you can add/remove stops in the demo application. its pretty fun to play around with different settings.

next step is to update the options based on the "from CSS gradient" input field.

**Comment by vikpe** [comment]
Now it tries to apply options from CSS gradient. The lil-gui input is very small so typing manually is somewhat tedious, pasting values works fine.

[css_gradient.webm](https://user-images.githubusercontent.com/1616817/209737021-b4efc6bb-1642-4104-b611-eba1fc0f6134.webm)

**Comment by vikpe** [comment]
@bigtimebuddy the filter is pretty much ready to be published, however I'd like to utilize the color class (pixijs/pixijs#9061), so I'll wait a bit.

**Comment by vikpe** [comment]
It's more of a nice to have, we can publish and collect some feedback.

**Review by bigtimebuddy — APPROVED** [review]

**Comment by bigtimebuddy** [comment]
Could you resolve these conflicts?

**Pull request merged**